File: src/main/java/tarehart/rlbot/ui/PlainReadout.java
package tarehart.rlbot.ui;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.BallSlice;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.BallPath;
import tarehart.rlbot.planning.*;
import tarehart.rlbot.time.Duration;
import tarehart.rlbot.time.GameTime;
import tarehart.rlbot.tuning.BallPrediction;
import tarehart.rlbot.tuning.PredictionWarehouse;

import javax.swing.*;
import javax.swing.text.DefaultCaret;
import java.awt.*;
import java.util.Optional;

public class PlainReadout {

    private static final int HEIGHT_BAR_MULTIPLIER = 10;

    private JLabel planPosture;
    private JProgressBar ballHeightActual;
    private JProgressBar ballHeightPredicted;
    private JSlider predictionTime;
    private JPanel rootPanel;
    private JProgressBar ballHeightActualMax;
    private JProgressBar ballHeightPredictedMax;
    private JLabel situationText;
    private JLabel predictionTimeSeconds;
    private JTextArea logViewer;
    private JLabel ownGoalFutureProximity;
    private JLabel distanceBallIsBehindUs;
    private JLabel enemyOffensiveApproachError;
    private JLabel distanceFromEnemyBackWall;
    private JLabel distanceFromEnemyCorner;
    private JLabel needsDefensiveClear;
    private JLabel shotOnGoalAvailable;
    private JLabel expectedEnemyContact;
    private JLabel scoredOnThreat;
    private ArenaDisplay arenaDisplay;
    private JLabel goForKickoff;
    private JLabel waitToClear;
    private JLabel forceDefensivePosture;
    private JTextPane omniText;
    private JTextField logFilter;

    private double maxCarSpeedVal;

    private GameTime actualMaxTime = new GameTime(0);
    private GameTime predictedMaxTime = new GameTime(0);
    private PredictionWarehouse warehouse = new PredictionWarehouse();
    private GameTime previousTime = null;


    public PlainReadout() {
        setupUi();
        DefaultCaret caret = (DefaultCaret) logViewer.getCaret();
        caret.setUpdatePolicy(DefaultCaret.ALWAYS_UPDATE);
    }

    public void update(AgentInput input, Plan.Posture posture, String situation, String log, BallPath ballPath) {

        planPosture.setText(posture.name());
        situationText.setText(situation);
        predictionTimeSeconds.setText(String.format("%.2f", predictionTime.getValue() / 1000.0));
        // ballHeightPredicted.setValue(0); // Commented out to avoid flicker. Should always be fresh anyway.
        ballHeightActual.setValue((int) (input.getBallPosition().getZ() * HEIGHT_BAR_MULTIPLIER));

        // Filter log before appending
        String filterValue = logFilter.getText();
        if (filterValue != null && filterValue.length() > 0) {
            StringBuilder newLog = new StringBuilder();
            filterValue = filterValue.toLowerCase();
            String[] splitLog = log.split("\n");

            for (int i = 0; i < splitLog.length; i++) {
                if (splitLog[i].toLowerCase().contains(filterValue)) {
                    newLog.append(splitLog[i] + "\n");
                }
            }

            log = newLog.toString();
        }
        logViewer.append(log);

        gatherBallPredictionData(input, ballPath);
        Optional<Vector3> ballPredictionOption = getBallPrediction(input);
        if (ballPredictionOption.isPresent()) {
            Vector3 ballPrediction = ballPredictionOption.get();
            ballHeightPredicted.setValue((int) (ballPrediction.getZ() * HEIGHT_BAR_MULTIPLIER));
            arenaDisplay.updateBallPrediction(ballPrediction);
        }

        TacticalSituation tacSituation = TacticsTelemetry.INSTANCE.get(input.getPlayerIndex());

        if (tacSituation != null) {
            arenaDisplay.updateExpectedEnemyContact(Optional.ofNullable(tacSituation.getExpectedEnemyContact()));
        }

        updateBallHeightMaxes(input);
        updateTacticsInfo(input);
        updateOmniText(input, Optional.ofNullable(tacSituation));
        arenaDisplay.updateInput(input);
        arenaDisplay.repaint();
    }

    private void updateBallHeightMaxes(AgentInput input) {
        // Calculate and display Ball Height Actual Max
        if (ballHeightActualMax.getValue() < ballHeightActual.getValue()) {
            ballHeightActualMax.setValue(ballHeightActual.getValue());
            actualMaxTime = input.getTime();
        } else if (Duration.Companion.between(input.getTime(), actualMaxTime).abs().getSeconds() > 3) {
            ballHeightActualMax.setValue(0);
        }

        // Calculate and display Ball Height Predicted Max
        if (ballHeightPredictedMax.getValue() < ballHeightPredicted.getValue()) {
            ballHeightPredictedMax.setValue(ballHeightPredicted.getValue());
            predictedMaxTime = input.getTime();
        } else if (Duration.Companion.between(input.getTime(), predictedMaxTime).abs().getSeconds() > 3) {
            ballHeightPredictedMax.setValue(0);
        }
    }

    private void gatherBallPredictionData(AgentInput input, BallPath ballPath) {
        int predictionMillis = predictionTime.getValue();
        GameTime predictionTime = input.getTime().plus(Duration.Companion.ofMillis(predictionMillis));

        if (previousTime == null || !previousTime.equals(input.getTime())) {
            if (ballPath != null) {
                BallSlice predictionSpace = ballPath.getMotionAt(predictionTime);
                if (predictionSpace != null) {
                    BallPrediction prediction = new BallPrediction(predictionSpace.getSpace(), predictionTime, ballPath);
                    warehouse.addPrediction(prediction);
                }
                previousTime = input.getTime();
            }
        }
    }

    private Optional<Vector3> getBallPrediction(AgentInput input) {
        Optional<BallPrediction> predictionOfNow = warehouse.getPredictionOfMoment(input.getTime());
        return predictionOfNow.map(ballPrediction -> ballPrediction.predictedLocation);
    }

    private void updateOmniText(AgentInput input, Optional<TacticalSituation> situation) {

        ZonePlan zonePlan = ZoneTelemetry.INSTANCE.get(input.getPlayerIndex());

        Optional<CarData> enemy = situation.map(TacticalSituation::getEnemyPlayerWithInitiative).map(CarWithIntercept::getCar);

        String text = "" +
                "Our Pos: " + input.getMyCarData().getPosition() + "\n" +
                "Our Vel: " + input.getMyCarData().getVelocity() + "\n" +
                "Enemy Pos: " + enemy.map(CarData::getPosition).orElse(new Vector3()) + "\n" +
                "Enemy Vel: " + enemy.map(CarData::getVelocity).orElse(new Vector3()) + "\n" +
                "Ball Pos: " + input.getBallPosition() + "\n" +
                "\n" +
                "Our Car Zone: " + (zonePlan != null ? printCarZone(zonePlan) : "Unknown") + "\n" +
                "Ball Zone: " + (zonePlan != null ? zonePlan.getBallZone().toString() : "Unknown") + "\n" +
                "\n" +
                "Ball Spin: " + input.getBallSpin() + "\n" +
                "\n" +
                "Our Boost: " + input.getMyCarData().getBoost();

        this.omniText.setText(text);
    }

    private String printCarZone(ZonePlan zp) {
        Zone zone = zp.getMyZone();
        return zone.getMainZone() + " " + zone.getSubZone();
    }

    private void updateTacticsInfo(AgentInput input) {
        TacticalSituation situation = TacticsTelemetry.INSTANCE.get(input.getPlayerIndex());

        if (situation != null) {

            ownGoalFutureProximity.setText(String.format("%.2f", situation.getOwnGoalFutureProximity()));
            distanceBallIsBehindUs.setText(String.format("%.2f", situation.getDistanceBallIsBehindUs()));
            enemyOffensiveApproachError.setText(Optional.ofNullable(situation.getEnemyOffensiveApproachError())
                    .map(e -> String.format("%.2f", e)).orElse(""));
            distanceFromEnemyBackWall.setText(String.format("%.2f", situation.getDistanceFromEnemyBackWall()));
            distanceFromEnemyCorner.setText(String.format("%.2f", situation.getDistanceFromEnemyCorner()));
            expectedEnemyContact.setText(Optional.ofNullable(situation.getExpectedEnemyContact())
                    .map(contact -> contact.getSpace().toString()).orElse(""));
            scoredOnThreat.setText(Optional.ofNullable(situation.getScoredOnThreat())
                    .map(b -> b.getSpace().toString()).orElse("None"));

            needsDefensiveClear.setText("");
            needsDefensiveClear.setBackground(situation.getNeedsDefensiveClear() ? Color.GREEN : Color.WHITE);
            needsDefensiveClear.setOpaque(situation.getNeedsDefensiveClear());

            shotOnGoalAvailable.setText("");
            shotOnGoalAvailable.setBackground(situation.getShotOnGoalAvailable() ? Color.GREEN : Color.WHITE);
            shotOnGoalAvailable.setOpaque(situation.getShotOnGoalAvailable());

            goForKickoff.setText("");
            goForKickoff.setBackground(situation.getGoForKickoff() ? Color.GREEN : Color.WHITE);
            goForKickoff.setOpaque(situation.getGoForKickoff());

            waitToClear.setText("");
            waitToClear.setBackground(situation.getWaitToClear() ? Color.GREEN : Color.WHITE);
            waitToClear.setOpaque(situation.getWaitToClear());

            forceDefensivePosture.setText("");
            forceDefensivePosture.setBackground(situation.getForceDefensivePosture() ? Color.GREEN : Color.WHITE);
            forceDefensivePosture.setOpaque(situation.getForceDefensivePosture());
        }
    }

    public JPanel getRootPanel() {
        return rootPanel;
    }

    /**
     * @noinspection ALL
     */
    private Font getFont(String fontName, int style, int size, Font currentFont) {
        if (currentFont == null) return null;
        String resultName;
        if (fontName == null) {
            resultName = currentFont.getName();
        } else {
            Font testFont = new Font(fontName, Font.PLAIN, 10);
            if (testFont.canDisplay('a') && testFont.canDisplay('1')) {
                resultName = fontName;
            } else {
                resultName = currentFont.getName();
            }
        }
        return new Font(resultName, style >= 0 ? style : currentFont.getStyle(), size >= 0 ? size : currentFont.getSize());
    }

    {
        setupUi();
    }

    private void setupUi() {
        rootPanel = new JPanel(new BorderLayout());
        rootPanel.setMinimumSize(new Dimension(400, 600));

        situationText = new JLabel();
        JPanel situationPanel = new JPanel();
        situationPanel.add(situationText);
        rootPanel.add(situationPanel, BorderLayout.NORTH);

        arenaDisplay = new ArenaDisplay();
        arenaDisplay.setBackground(new Color(0xFCFCFC));
        rootPanel.add(arenaDisplay, BorderLayout.CENTER);
        final JPanel diagnostics = new JPanel();
        diagnostics.setLayout(new BoxLayout(diagnostics, BoxLayout.Y_AXIS));
        rootPanel.add(diagnostics, BorderLayout.EAST);
        final JScrollPane scrollPane = new JScrollPane();
        diagnostics.add(scrollPane);
        logViewer = new JTextArea();
        logViewer.setEditable(false);
        scrollPane.setViewportView(logViewer);
        scrollPane.setMaximumSize(new Dimension(400, 600));
        scrollPane.setPreferredSize(new Dimension(400, 600));
        diagnostics.add(new JLabel("Log Filter"));
        logFilter = new JTextField();
        diagnostics.add(logFilter);
        final JPanel extras = new JPanel();
        extras.setLayout(new GridLayout(0, 2));
        diagnostics.add(extras);
        extras.add(new JLabel("Posture"));
        planPosture = new JLabel("NEUTRAL");
        extras.add(planPosture);
        extras.add(new JLabel("Ball Height Actual"));
        ballHeightActual = new JProgressBar();
        ballHeightActual.setMaximum(450);
        extras.add(ballHeightActual);

        extras.add(new JLabel("Ball Height A. Max"));
        ballHeightActualMax = new JProgressBar();
        ballHeightActualMax.setMaximum(450);
        extras.add(ballHeightActualMax);

        extras.add(new JLabel("Ball Height Predicted"));
        ballHeightPredicted = new JProgressBar();
        ballHeightPredicted.setMaximum(450);
        extras.add(ballHeightPredicted);

        extras.add(new JLabel("Ball Height P. Max"));
        ballHeightPredictedMax = new JProgressBar();
        ballHeightPredictedMax.setMaximum(450);
        extras.add(ballHeightPredictedMax);

        extras.add(new JLabel("Prediction Time (s)"));
        final JPanel predictionSliderPanel = new JPanel(new FlowLayout());
        extras.add(predictionSliderPanel);
        predictionTime = new JSlider();
        predictionTime.setMaximum(5000);
        predictionTime.setPaintTicks(true);
        predictionTime.setValue(1000);
        predictionSliderPanel.add(predictionTime);
        predictionTimeSeconds = new JLabel();
        predictionTimeSeconds.setText("1.000000");
        predictionSliderPanel.add(predictionTimeSeconds);



        final JPanel panel4 = new JPanel();
        panel4.setLayout(new FlowLayout());
        diagnostics.add(panel4);
        omniText = new JTextPane();
        omniText.setBackground(new Color(-1));
        Font omniTextFont = this.getFont("Monospaced", -1, -1, omniText.getFont());
        if (omniTextFont != null) omniText.setFont(omniTextFont);
        //panel4.add(omniText);
        final JPanel panel5 = new JPanel();
        panel5.setLayout(new GridLayout(0, 2));
        rootPanel.add(panel5, BorderLayout.WEST);
        final JLabel label9 = new JLabel();
        label9.setText("ownGoalFutureProximity");
        panel5.add(label9);
        ownGoalFutureProximity = new JLabel();
        ownGoalFutureProximity.setText("?");
        panel5.add(ownGoalFutureProximity);
        final JLabel label10 = new JLabel();
        label10.setText("distanceBallIsBehindUs");
        panel5.add(label10);
        distanceBallIsBehindUs = new JLabel();
        distanceBallIsBehindUs.setText("?");
        panel5.add(distanceBallIsBehindUs);
        final JLabel label11 = new JLabel();
        label11.setText("enemyOffensiveApproachError");
        panel5.add(label11);
        enemyOffensiveApproachError = new JLabel();
        enemyOffensiveApproachError.setText("?");
        panel5.add(enemyOffensiveApproachError);
        final JLabel label12 = new JLabel();
        label12.setText("distanceFromEnemyBackWall");
        panel5.add(label12);
        distanceFromEnemyBackWall = new JLabel();
        distanceFromEnemyBackWall.setText("?");
        panel5.add(distanceFromEnemyBackWall);
        final JLabel label13 = new JLabel();
        label13.setText("expectedEnemyContact");
        panel5.add(label13);
        expectedEnemyContact = new JLabel();
        expectedEnemyContact.setText("?");
        panel5.add(expectedEnemyContact);
        final JLabel label14 = new JLabel();
        label14.setText("scoredOnThreat");
        panel5.add(label14);
        scoredOnThreat = new JLabel();
        scoredOnThreat.setText("?");
        panel5.add(scoredOnThreat);

        final JLabel label15 = new JLabel();
        label15.setText("distanceFromEnemyCorner");
        panel5.add(label15);
        distanceFromEnemyCorner = new JLabel();
        distanceFromEnemyCorner.setText("?");
        panel5.add(distanceFromEnemyCorner);

        panel5.add(new JLabel("waitToClear"));
        waitToClear = new JLabel();
        waitToClear.setText("?");
        panel5.add(waitToClear);

        panel5.add(new JLabel("forceDefensivePosture"));
        forceDefensivePosture = new JLabel();
        forceDefensivePosture.setText("?");
        panel5.add(forceDefensivePosture);

        panel5.add(new JLabel("shotOnGoalAvailable"));
        shotOnGoalAvailable = new JLabel();
        shotOnGoalAvailable.setText("?");
        panel5.add(shotOnGoalAvailable);

        panel5.add(new JLabel("needsDefensiveClear"));
        needsDefensiveClear = new JLabel();
        needsDefensiveClear.setText("?");
        panel5.add(needsDefensiveClear);

        panel5.add(new JLabel("goForKickoff"));
        goForKickoff = new JLabel();
        goForKickoff.setText("?");
        panel5.add(goForKickoff);
    }

    /**
     * @noinspection ALL
     */
    public JComponent getRootComponent() {
        return rootPanel;
    }

}

File: src/main/java/tarehart/rlbot/ui/PlainReadout.kt
package tarehart.rlbot.ui

import sun.awt.SunToolkit
import tarehart.rlbot.AgentInput
import tarehart.rlbot.intercept.Intercept
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.physics.BallPath
import tarehart.rlbot.planning.*
import tarehart.rlbot.time.Duration
import tarehart.rlbot.time.GameTime
import tarehart.rlbot.tuning.BallPrediction
import tarehart.rlbot.tuning.PredictionWarehouse
import java.awt.*
import java.util.*
import javax.swing.*
import javax.swing.text.DefaultCaret

class PlainReadout {

    private val planPosture = JLabel("NEUTRAL")
    private val ballHeightActual = JProgressBar()
    private val ballHeightPredicted = JProgressBar()
    private val predictionTime = JSlider()
    val rootPanel = JPanel(BorderLayout())
    private val ballHeightActualMax = JProgressBar()
    private val ballHeightPredictedMax = JProgressBar()
    private val situationText = JLabel()
    private val predictionTimeSeconds = JLabel()
    private val logViewer = JTextArea()
    private val ownGoalFutureProximity = JLabel()
    private val distanceBallIsBehindUs = JLabel()
    private val enemyOffensiveApproachError = JLabel()
    private val distanceFromEnemyBackWall = JLabel()
    private val distanceFromEnemyCorner = JLabel()
    private val needsDefensiveClear = JLabel()
    private val shotOnGoalAvailable = JLabel()
    private val expectedEnemyContact = JLabel()
    private val scoredOnThreat = JLabel()
    private val arenaDisplay = ArenaDisplay()
    private val goForKickoff = JLabel()
    private val waitToClear = JLabel()
    private val forceDefensivePosture = JLabel()
    private val omniText = JTextPane()
    private val logFilter = JTextField()

    private var actualMaxTime = GameTime(0)
    private var predictedMaxTime = GameTime(0)
    private val warehouse = PredictionWarehouse()
    private var previousTime: GameTime? = null

    init {
        setupUi()
        val caret = logViewer.caret as DefaultCaret
        caret.updatePolicy = DefaultCaret.ALWAYS_UPDATE
    }

    fun update(input: AgentInput, posture: Plan.Posture, situation: String, log: String, ballPath: BallPath) {

        var log = log
        
        if (!SunToolkit.getContainingWindow(rootPanel).isVisible) {
            return
        }

        planPosture.text = posture.name
        situationText.text = situation
        predictionTimeSeconds.text = String.format("%.2f", predictionTime.value / 1000.0)
        // ballHeightPredicted.setValue(0); // Commented out to avoid flicker. Should always be fresh anyway.
        ballHeightActual.value = (input.ballPosition.z * HEIGHT_BAR_MULTIPLIER).toInt()

        // Filter log before appending
        var filterValue: String? = logFilter.text
        if (filterValue != null && filterValue.isNotEmpty()) {
            val newLog = StringBuilder()
            filterValue = filterValue.toLowerCase()
            val splitLog = log.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            for (i in splitLog.indices) {
                if (splitLog[i].toLowerCase().contains(filterValue)) {
                    newLog.append(splitLog[i] + "\n")
                }
            }

            log = newLog.toString()
        }
        logViewer.append(log)

        gatherBallPredictionData(input, ballPath)
        val ballPredictionOption = getBallPrediction(input)
        if (ballPredictionOption.isPresent) {
            val ballPrediction = ballPredictionOption.get()
            ballHeightPredicted.value = (ballPrediction.z * HEIGHT_BAR_MULTIPLIER).toInt()
            arenaDisplay.updateBallPrediction(ballPrediction)
        }

        val tacSituation = TacticsTelemetry[input.playerIndex]

        if (tacSituation != null) {
            arenaDisplay.updateExpectedEnemyContact(tacSituation.expectedEnemyContact)
        }

        updateBallHeightMaxes(input)
        updateTacticsInfo(input)
        updateOmniText(input, tacSituation)
        arenaDisplay.updateInput(input)
        arenaDisplay.repaint()
    }

    private fun updateBallHeightMaxes(input: AgentInput) {
        // Calculate and display Ball Height Actual Max
        if (ballHeightActualMax.value < ballHeightActual.value) {
            ballHeightActualMax.value = ballHeightActual.value
            actualMaxTime = input.time
        } else if (Duration.between(input.time, actualMaxTime).abs().seconds > 3) {
            ballHeightActualMax.value = 0
        }

        // Calculate and display Ball Height Predicted Max
        if (ballHeightPredictedMax.value < ballHeightPredicted.value) {
            ballHeightPredictedMax.value = ballHeightPredicted.value
            predictedMaxTime = input.time
        } else if (Duration.between(input.time, predictedMaxTime).abs().seconds > 3) {
            ballHeightPredictedMax.value = 0
        }
    }

    private fun gatherBallPredictionData(input: AgentInput, ballPath: BallPath?) {
        val predictionMillis = predictionTime.value
        val predictionTime = input.time.plus(Duration.ofMillis(predictionMillis.toLong()))

        if (previousTime == null || previousTime != input.time) {
            if (ballPath != null) {
                val predictionSpace = ballPath.getMotionAt(predictionTime)
                if (predictionSpace != null) {
                    val prediction = BallPrediction(predictionSpace.space, predictionTime, ballPath)
                    warehouse.addPrediction(prediction)
                }
                previousTime = input.time
            }
        }
    }

    private fun getBallPrediction(input: AgentInput): Optional<Vector3> {
        val predictionOfNow = warehouse.getPredictionOfMoment(input.time)
        return predictionOfNow.map { ballPrediction -> ballPrediction.predictedLocation }
    }

    private fun updateOmniText(input: AgentInput, situation: TacticalSituation?) {

        val zonePlan = ZoneTelemetry[input.playerIndex]

        val enemy = situation?.enemyPlayerWithInitiative?.car

        val text = "" +
                "Our Pos: " + input.myCarData.position + "\n" +
                "Our Vel: " + input.myCarData.velocity + "\n" +
                "Enemy Pos: " + (enemy?.position ?: Vector3()) + "\n" +
                "Enemy Vel: " + (enemy?.velocity ?: Vector3()) + "\n" +
                "Ball Pos: " + input.ballPosition + "\n" +
                "\n" +
                "Our Car Zone: " + (if (zonePlan != null) printCarZone(zonePlan) else "Unknown") + "\n" +
                "Ball Zone: " + (zonePlan?.ballZone?.toString() ?: "Unknown") + "\n" +
                "\n" +
                "Ball Spin: " + input.ballSpin + "\n" +
                "\n" +
                "Our Boost: " + input.myCarData.boost

        this.omniText.text = text
    }

    private fun printCarZone(zp: ZonePlan): String {
        val zone = zp.myZone
        return zone.mainZone.toString() + " " + zone.subZone
    }

    private fun updateTacticsInfo(input: AgentInput) {
        val situation = TacticsTelemetry[input.playerIndex]

        if (situation != null) {

            ownGoalFutureProximity.text = String.format("%.2f", situation.ownGoalFutureProximity)
            distanceBallIsBehindUs.text = String.format("%.2f", situation.distanceBallIsBehindUs)
            enemyOffensiveApproachError.text = Optional.ofNullable(situation.enemyOffensiveApproachError)
                    .map { e -> String.format("%.2f", e) }.orElse("")
            distanceFromEnemyBackWall.text = String.format("%.2f", situation.distanceFromEnemyBackWall)
            distanceFromEnemyCorner.text = String.format("%.2f", situation.distanceFromEnemyCorner)
            expectedEnemyContact.text = Optional.ofNullable<Intercept>(situation.expectedEnemyContact)
                    .map<String> { contact -> contact.space.toString() }.orElse("")
            scoredOnThreat.text = Optional.ofNullable(situation.scoredOnThreat)
                    .map { (space) -> space.toString() }.orElse("None")

            needsDefensiveClear.text = ""
            needsDefensiveClear.background = if (situation.needsDefensiveClear) Color.GREEN else Color.WHITE
            needsDefensiveClear.isOpaque = situation.needsDefensiveClear

            shotOnGoalAvailable.text = ""
            shotOnGoalAvailable.background = if (situation.shotOnGoalAvailable) Color.GREEN else Color.WHITE
            shotOnGoalAvailable.isOpaque = situation.shotOnGoalAvailable

            goForKickoff.text = ""
            goForKickoff.background = if (situation.goForKickoff) Color.GREEN else Color.WHITE
            goForKickoff.isOpaque = situation.goForKickoff

            waitToClear.text = ""
            waitToClear.background = if (situation.waitToClear) Color.GREEN else Color.WHITE
            waitToClear.isOpaque = situation.waitToClear

            forceDefensivePosture.text = ""
            forceDefensivePosture.background = if (situation.forceDefensivePosture) Color.GREEN else Color.WHITE
            forceDefensivePosture.isOpaque = situation.forceDefensivePosture
        }
    }

    /**
     * @noinspection ALL
     */
    private fun getFont(fontName: String?, style: Int, size: Int, currentFont: Font?): Font? {
        if (currentFont == null) return null
        val resultName: String
        if (fontName == null) {
            resultName = currentFont.name
        } else {
            val testFont = Font(fontName, Font.PLAIN, 10)
            if (testFont.canDisplay('a') && testFont.canDisplay('1')) {
                resultName = fontName
            } else {
                resultName = currentFont.name
            }
        }
        return Font(resultName, if (style >= 0) style else currentFont.style, if (size >= 0) size else currentFont.size)
    }

    init {
        setupUi()
    }

    private fun setupUi() {
        rootPanel.minimumSize = Dimension(400, 600)

        val situationPanel = JPanel()
        situationPanel.add(situationText)
        rootPanel.add(situationPanel, BorderLayout.NORTH)

        arenaDisplay.background = Color(0xFCFCFC)
        rootPanel.add(arenaDisplay, BorderLayout.CENTER)
        val diagnostics = JPanel()
        diagnostics.layout = BoxLayout(diagnostics, BoxLayout.Y_AXIS)
        rootPanel.add(diagnostics, BorderLayout.EAST)
        val scrollPane = JScrollPane()
        diagnostics.add(scrollPane)
        logViewer.isEditable = false
        scrollPane.setViewportView(logViewer)
        scrollPane.maximumSize = Dimension(400, 600)
        scrollPane.preferredSize = Dimension(400, 600)
        diagnostics.add(JLabel("Log Filter"))
        diagnostics.add(logFilter)
        val extras = JPanel()
        extras.layout = GridLayout(0, 2)
        diagnostics.add(extras)
        extras.add(JLabel("Posture"))
        extras.add(planPosture)
        extras.add(JLabel("Ball Height Actual"))
        ballHeightActual.maximum = 450
        extras.add(ballHeightActual)

        extras.add(JLabel("Ball Height A. Max"))
        ballHeightActualMax.maximum = 450
        extras.add(ballHeightActualMax)

        extras.add(JLabel("Ball Height Predicted"))
        ballHeightPredicted.maximum = 450
        extras.add(ballHeightPredicted)

        extras.add(JLabel("Ball Height P. Max"))
        ballHeightPredictedMax.maximum = 450
        extras.add(ballHeightPredictedMax)

        extras.add(JLabel("Prediction Time (s)"))
        val predictionSliderPanel = JPanel(FlowLayout())
        extras.add(predictionSliderPanel)
        predictionTime.maximum = 5000
        predictionTime.paintTicks = true
        predictionTime.value = 1000
        predictionSliderPanel.add(predictionTime)
        predictionTimeSeconds.text = "1.000000"
        predictionSliderPanel.add(predictionTimeSeconds)


        val panel4 = JPanel()
        panel4.layout = FlowLayout()
        diagnostics.add(panel4)
        omniText.background = Color(-1)
        val omniTextFont = this.getFont("Monospaced", -1, -1, omniText.font)
        if (omniTextFont != null) omniText.font = omniTextFont
        //panel4.add(omniText);
        val panel5 = JPanel()
        panel5.layout = GridLayout(0, 2)
        rootPanel.add(panel5, BorderLayout.WEST)
        val label9 = JLabel()
        label9.text = "ownGoalFutureProximity"
        panel5.add(label9)
        ownGoalFutureProximity.text = "?"
        panel5.add(ownGoalFutureProximity)
        val label10 = JLabel()
        label10.text = "distanceBallIsBehindUs"
        panel5.add(label10)
        distanceBallIsBehindUs.text = "?"
        panel5.add(distanceBallIsBehindUs)
        val label11 = JLabel()
        label11.text = "enemyOffensiveApproachError"
        panel5.add(label11)
        enemyOffensiveApproachError.text = "?"
        panel5.add(enemyOffensiveApproachError)
        val label12 = JLabel()
        label12.text = "distanceFromEnemyBackWall"
        panel5.add(label12)
        distanceFromEnemyBackWall.text = "?"
        panel5.add(distanceFromEnemyBackWall)
        val label13 = JLabel()
        label13.text = "expectedEnemyContact"
        panel5.add(label13)
        expectedEnemyContact.text = "?"
        panel5.add(expectedEnemyContact)
        val label14 = JLabel()
        label14.text = "scoredOnThreat"
        panel5.add(label14)
        scoredOnThreat.text = "?"
        panel5.add(scoredOnThreat)

        val label15 = JLabel()
        label15.text = "distanceFromEnemyCorner"
        panel5.add(label15)
        distanceFromEnemyCorner.text = "?"
        panel5.add(distanceFromEnemyCorner)

        panel5.add(JLabel("waitToClear"))
        waitToClear.text = "?"
        panel5.add(waitToClear)

        panel5.add(JLabel("forceDefensivePosture"))
        forceDefensivePosture.text = "?"
        panel5.add(forceDefensivePosture)

        panel5.add(JLabel("shotOnGoalAvailable"))
        shotOnGoalAvailable.text = "?"
        panel5.add(shotOnGoalAvailable)

        panel5.add(JLabel("needsDefensiveClear"))
        needsDefensiveClear.text = "?"
        panel5.add(needsDefensiveClear)

        panel5.add(JLabel("goForKickoff"))
        goForKickoff.text = "?"
        panel5.add(goForKickoff)
    }

    companion object {

        private const val HEIGHT_BAR_MULTIPLIER = 10
    }

}
